File: FlightRadar24/flight.py
# -*- coding: utf-8 -*-

class Flight(object):

    __default_text = "N/A"

    def __init__(self, flight_id, info):

        self.id = flight_id
        self.icao_24bit = self.__get_info(info[0])
        self.latitude = self.__get_info(info[1])
        self.longitude = self.__get_info(info[2])
        self.heading = self.__get_info(info[3])
        self.altitude = self.__get_info(info[4])
        self.ground_speed = self.__get_info(info[5])
        self.squawk = self.__get_info(info[6])
        self.aircraft_code = self.__get_info(info[8])
        self.registration = self.__get_info(info[9])
        self.time = self.__get_info(info[10])
        self.origin_airport_iata = self.__get_info(info[11])
        self.destination_airport_iata = self.__get_info(info[12])
        self.number = self.__get_info(info[13])
        self.airline_iata = self.__get_info(info[13][:2])
        self.on_ground = self.__get_info(info[14])
        self.vertical_speed =self.__get_info(info[15])
        self.callsign = self.__get_info(info[16])
        self.airline_icao = self.__get_info(info[18])

    def __get_details(self, data):
        return dict() if data is None else data

    def __get_info(self, info):

        return info if (info or info == 0) and info != self.__default_text else self.__default_text

    def __repr__(self):

        return self.__str__()

    def __str__(self):

        template = "<({}) {} - Altitude: {} - Ground Speed: {} - Heading: {}>"
        return template.format(self.aircraft_code, self.registration, self.altitude, self.ground_speed, self.heading)

    def check_info(self, **info):

        """
        Check one or more flight information.

        You can use the prefix "max_" or "min_" in the parameter
        to compare numeric data with ">" or "<".

        Example: check_info(min_altitude = 6700, max_altitude = 13000, icao = "THY")
        """

        comparison_functions = {"max": max, "min": min}

        for key, value in info.items():

            # Separate the comparison prefix if it exists.
            prefix, key = key.split("_", maxsplit = 1) if key[:4] == "max_" or key[:4] == "min_" else (None, key)

            # Check if the value is greater than or less than the attribute value.
            if prefix and key in self.__dict__:
                if comparison_functions[prefix](value, self.__dict__[key]) != value: return False

            # Check if the values ​​are equal.
            elif key in self.__dict__ and value != self.__dict__[key]: return False

        return True

    def get_altitude(self):

        return "{} ft".format(self.altitude)

    def get_flight_level(self):

        return str(self.altitude)[:3] + " FL" if self.altitude > 10000 else self.get_altitude()

    def get_ground_speed(self):

        return "{} kt".format(self.ground_speed) + ("s" if self.ground_speed > 1 else "")

    def get_heading(self):

        return str(self.heading) + "°"

    def get_vertical_speed(self):

        return "{} fpm".format(self.vertical_speed)

    def set_flight_details(self, flight_details):

        # Get aircraft data.
        aircraft = self.__get_details(flight_details.get("aircraft"))

        # Get airline data.
        airline = self.__get_details(flight_details.get("airline"))

        # Get airport data.
        airport = self.__get_details(flight_details.get("airport"))

        # Get destination data.
        dest_aiport = self.__get_details(airport.get("destination"))
        dest_aiport_code = self.__get_details(dest_aiport.get("code"))
        dest_aiport_info = self.__get_details(dest_aiport.get("info"))
        dest_aiport_position = self.__get_details(dest_aiport.get("position"))
        dest_aiport_country = self.__get_details(dest_aiport_position.get("country"))
        dest_aiport_timezone = self.__get_details(dest_aiport.get("timezone"))

        # Get origin data.
        orig_aiport = self.__get_details(airport.get("origin"))
        orig_aiport_code = self.__get_details(orig_aiport.get("code"))
        orig_aiport_info = self.__get_details(orig_aiport.get("info"))
        orig_aiport_position = self.__get_details(orig_aiport.get("position"))
        orig_aiport_country = self.__get_details(orig_aiport_position.get("country"))
        orig_aiport_timezone = self.__get_details(orig_aiport.get("timezone"))

        # Get flight history.
        history = self.__get_details(flight_details.get("flightHistory"))

        # Get flight status.
        status = self.__get_details(flight_details.get("status"))

        # Aircraft information.
        self.aircraft_age = self.__get_info(aircraft.get("age"))
        self.aircraft_country_id = self.__get_info(aircraft.get("countryId"))
        self.aircraft_history = history.get("aircraft", list())
        self.aircraft_images = aircraft.get("images", list())
        self.aircraft_model = self.__get_info(self.__get_details(aircraft.get("model")).get("text"))

        # Airline information.
        self.airline_name = self.__get_info(airline.get("name"))
        self.airline_short_name = self.__get_info(airline.get("short"))

        # Destination airport position.
        self.destination_airport_altitude = self.__get_info(dest_aiport_position.get("altitude"))
        self.destination_airport_country_code = self.__get_info(dest_aiport_country.get("code"))
        self.destination_airport_country_name = self.__get_info(dest_aiport_country.get("name"))
        self.destination_airport_latitude = self.__get_info(dest_aiport_position.get("latitude"))
        self.destination_airport_longitude = self.__get_info(dest_aiport_position.get("longitude"))

        # Destination airport information.
        self.destination_airport_icao = self.__get_info(dest_aiport_code.get("icao"))
        self.destination_airport_baggage = self.__get_info(dest_aiport_info.get("baggage"))
        self.destination_airport_gate = self.__get_info(dest_aiport_info.get("gate"))
        self.destination_airport_name = self.__get_info(dest_aiport.get("name"))
        self.destination_airport_terminal = self.__get_info(dest_aiport_info.get("terminal"))
        self.destination_airport_visible = self.__get_info(dest_aiport.get("visible"))
        self.destination_airport_website = self.__get_info(dest_aiport.get("website"))

        # Destination airport timezone.
        self.destination_airport_timezone_abbr = self.__get_info(dest_aiport_timezone.get("abbr"))
        self.destination_airport_timezone_abbr_name = self.__get_info(dest_aiport_timezone.get("abbrName"))
        self.destination_airport_timezone_name = self.__get_info(dest_aiport_timezone.get("name"))
        self.destination_airport_timezone_offset = self.__get_info(dest_aiport_timezone.get("offset"))
        self.destination_airport_timezone_offsetHours = self.__get_info(dest_aiport_timezone.get("offsetHours"))

        # Origin airport position.
        self.origin_airport_altitude = self.__get_info(orig_aiport_position.get("altitude"))
        self.origin_airport_country_code = self.__get_info(orig_aiport_country.get("code"))
        self.origin_airport_country_name = self.__get_info(orig_aiport_country.get("name"))
        self.origin_airport_latitude = self.__get_info(orig_aiport_position.get("latitude"))
        self.origin_airport_longitude = self.__get_info(orig_aiport_position.get("longitude"))

        # Origin airport information.
        self.origin_airport_icao = self.__get_info(orig_aiport_code.get("icao"))
        self.origin_airport_baggage = self.__get_info(orig_aiport_info.get("baggage"))
        self.origin_airport_gate = self.__get_info(orig_aiport_info.get("gate"))
        self.origin_airport_name = self.__get_info(orig_aiport.get("name"))
        self.origin_airport_terminal = self.__get_info(orig_aiport_info.get("terminal"))
        self.origin_airport_visible = self.__get_info(orig_aiport.get("visible"))
        self.origin_airport_website = self.__get_info(orig_aiport.get("website"))

        # Origin airport timezone.
        self.origin_airport_timezone_abbr = self.__get_info(orig_aiport_timezone.get("abbr"))
        self.origin_airport_timezone_abbr_name = self.__get_info(orig_aiport_timezone.get("abbrName"))
        self.origin_airport_timezone_name = self.__get_info(orig_aiport_timezone.get("name"))
        self.origin_airport_timezone_offset = self.__get_info(orig_aiport_timezone.get("offset"))
        self.origin_airport_timezone_offsetHours = self.__get_info(orig_aiport_timezone.get("offsetHours"))

        # Flight status.
        self.status_icon = self.__get_info(status.get("icon"))
        self.status_text = self.__get_info(status.get("text"))

        # Time details.
        self.time_details = self.__get_details(flight_details.get("time"))

        # Flight trail.
        self.trail = flight_details.get("trail", list())

File: nodejs/flightradarapi/entities/flight.js
const Entity = require("./entity");


class Flight extends Entity {
    /**
     * Flight representation.
     */
   
    static __default_text = "N/A";

    /**
     * Constructor of Flight class.
     * 
     * @param {*} flight_id - The flight ID specifically used by FlightRadar24
     * @param {*} info - Dictionary with received data from FlightRadar24
     */
    constructor(flight_id, info) {
        super();
        
        this.__set_position(this.__getInfo(info[1]), this.__getInfo(info[2]));

        this.id = flight_id;
        this.icao_24bit = this.__getInfo(info[0]);
        this.heading = this.__getInfo(info[3]);
        this.altitude = this.__getInfo(info[4]);
        this.ground_speed = this.__getInfo(info[5]);
        this.squawk = this.__getInfo(info[6]);
        this.aircraft_code = this.__getInfo(info[8]);
        this.registration = this.__getInfo(info[9]);
        this.time = this.__getInfo(info[10]);
        this.origin_airport_iata = this.__getInfo(info[11]);
        this.destination_airport_iata = this.__getInfo(info[12]);
        this.number = this.__getInfo(info[13]);
        this.airline_iata = this.__getInfo(info[13].slice(0, 2));
        this.on_ground = this.__getInfo(info[14]);
        this.vertical_speed =this.__getInfo(info[15]);
        this.callsign = this.__getInfo(info[16]);
        this.airline_icao = this.__getInfo(info[18]);
    }

    /**
     * Check one or more flight information.
     *
     * You can use the prefix "max_" or "min_" in the parameter
     * to compare numeric data with ">" or "<".
     *
     * Example: check_info({min_altitude: 6700}, {max_altitude: 13000}, {icao: "THY"})
     * @param {object} info 
     * @returns {boolean}
     */
    check_info(info) {

        const comparisonFunctions = {"max": Math.max, "min": Math.min};

        for (let key in info) {
            let prefix = key.slice(0, 4);
            let value = info[key];
            
            // Separate the comparison prefix if it exists.
            if ((prefix === "max_") || (prefix === "min_")) {
                key = key.split("_", 1)[1];  
                prefix = prefix.slice(0, -1);
            }
            else {
                prefix = null;
            }

            // Check if the value is greater than or less than the attribute value.
            if (this.hasOwnProperty(key) && prefix) {
                if (comparisonFunctions[prefix](value, this[key]) !== value) {
                    return false;
                }
            }
            // Check if the value is equal.
            else if (this.hasOwnProperty(key) && value !== this[key]) {
                return false;
            }
        }
        return true;
    }

    /**
     * Return the formatted altitude, with the unit of measure.
     * @returns {string}
     */
    get_altitude() {
        return this.altitude.toString() + " ft";
    }

    /**
     * Return the formatted flight level, with the unit of measure.
     * @returns {string}
     */
    get_flight_level() {
        if (this.altitude >= 10000) {
            return this.altitude.toString().slice(0, 3) + " FL";
        }
        return this.get_altitude();
    }

    /**
     * Return the formatted ground speed, with the unit of measure.
     * @returns {string}
     */
    get_ground_speed() {
        const sufix = this.ground_speed > 1 ? "s" : "";
        return this.ground_speed.toString() + " kt" + sufix;
    }

    /**
     * Return the formatted heading, with the unit of measure.
     * @returns {string}
     */
    get_heading() {
        return this.heading.toString() + "°";
    }

    /**
     * Return the formatted vertical speed, with the unit of measure.
     * @returns {string}
     */
    get_vertical_speed() {
        return this.vertical_speed.toString() + " fpm";
    }

    /**
     * Set flight details to the instance. Use FlightRadar24API.get_flight_details(...) method to get it.
     * @param {object} flight_details
     */
    set_flight_details(flight_details) {
        flight_details = this.__createGetterMethodFor(flight_details);

        // Get aircraft data.
        const aircraft = this.__getDetails(flight_details.get("aircraft"));

        // Get airline data.
        const airline = this.__getDetails(flight_details.get("airline"));

        // Get airport data.
        const airport = this.__getDetails(flight_details.get("airport"));

        // Get destination data.
        const dest_airport = this.__getDetails(airport.get("destination"));
        const dest_airport_code = this.__getDetails(dest_airport.get("code"));
        const dest_airport_info = this.__getDetails(dest_airport.get("info"));
        const dest_airport_position = this.__getDetails(dest_airport.get("position"));
        const dest_airport_country = this.__getDetails(dest_airport_position.get("country"));
        const dest_airport_timezone = this.__getDetails(dest_airport.get("timezone"));

        // Get origin data.
        const orig_airport = this.__getDetails(airport.get("origin"));
        const orig_airport_code = this.__getDetails(orig_airport.get("code"));
        const orig_airport_info = this.__getDetails(orig_airport.get("info"));
        const orig_airport_position = this.__getDetails(orig_airport.get("position"));
        const orig_airport_country = this.__getDetails(orig_airport_position.get("country"));
        const orig_airport_timezone = this.__getDetails(orig_airport.get("timezone"));

        // Get flight history.
        const history = this.__getDetails(flight_details.get("flightHistory"));

        // Get flight status.
        const status = this.__getDetails(flight_details.get("status"));

        // Aircraft information.
        this.aircraft_age = this.__getInfo(aircraft.get("age"));
        this.aircraft_country_id = this.__getInfo(aircraft.get("countryId"));
        this.aircraft_history = history.get("aircraft", list());
        this.aircraft_images = aircraft.get("images", list());
        this.aircraft_model = this.__getInfo(this.__getDetails(aircraft.get("model")).get("text"));

        // Airline information.
        this.airline_name = this.__getInfo(airline.get("name"));
        this.airline_short_name = this.__getInfo(airline.get("short"));

        // Destination airport position.
        this.destination_airport_altitude = this.__getInfo(dest_airport_position.get("altitude"));
        this.destination_airport_country_code = this.__getInfo(dest_airport_country.get("code"));
        this.destination_airport_country_name = this.__getInfo(dest_airport_country.get("name"));
        this.destination_airport_latitude = this.__getInfo(dest_airport_position.get("latitude"));
        this.destination_airport_longitude = this.__getInfo(dest_airport_position.get("longitude"));

        // Destination airport information.
        this.destination_airport_icao = this.__getInfo(dest_airport_code.get("icao"));
        this.destination_airport_baggage = this.__getInfo(dest_airport_info.get("baggage"));
        this.destination_airport_gate = this.__getInfo(dest_airport_info.get("gate"));
        this.destination_airport_name = this.__getInfo(dest_airport.get("name"));
        this.destination_airport_terminal = this.__getInfo(dest_airport_info.get("terminal"));
        this.destination_airport_visible = this.__getInfo(dest_airport.get("visible"));
        this.destination_airport_website = this.__getInfo(dest_airport.get("website"));

        // Destination airport timezone.
        this.destination_airport_timezone_abbr = this.__getInfo(dest_airport_timezone.get("abbr"));
        this.destination_airport_timezone_abbr_name = this.__getInfo(dest_airport_timezone.get("abbrName"));
        this.destination_airport_timezone_name = this.__getInfo(dest_airport_timezone.get("name"));
        this.destination_airport_timezone_offset = this.__getInfo(dest_airport_timezone.get("offset"));
        this.destination_airport_timezone_offsetHours = this.__getInfo(dest_airport_timezone.get("offsetHours"));

        // Origin airport position.
        this.origin_airport_altitude = this.__getInfo(orig_airport_position.get("altitude"));
        this.origin_airport_country_code = this.__getInfo(orig_airport_country.get("code"));
        this.origin_airport_country_name = this.__getInfo(orig_airport_country.get("name"));
        this.origin_airport_latitude = this.__getInfo(orig_airport_position.get("latitude"));
        this.origin_airport_longitude = this.__getInfo(orig_airport_position.get("longitude"));

        // Origin airport information.
        this.origin_airport_icao = this.__getInfo(orig_airport_code.get("icao"));
        this.origin_airport_baggage = this.__getInfo(orig_airport_info.get("baggage"));
        this.origin_airport_gate = this.__getInfo(orig_airport_info.get("gate"));
        this.origin_airport_name = this.__getInfo(orig_airport.get("name"));
        this.origin_airport_terminal = this.__getInfo(orig_airport_info.get("terminal"));
        this.origin_airport_visible = this.__getInfo(orig_airport.get("visible"));
        this.origin_airport_website = this.__getInfo(orig_airport.get("website"));

        // Origin airport timezone.
        this.origin_airport_timezone_abbr = this.__getInfo(orig_airport_timezone.get("abbr"));
        this.origin_airport_timezone_abbr_name = this.__getInfo(orig_airport_timezone.get("abbrName"));
        this.origin_airport_timezone_name = this.__getInfo(orig_airport_timezone.get("name"));
        this.origin_airport_timezone_offset = this.__getInfo(orig_airport_timezone.get("offset"));
        this.origin_airport_timezone_offsetHours = this.__getInfo(orig_airport_timezone.get("offsetHours"));

        // Flight status.
        this.status_icon = this.__getInfo(status.get("icon"));
        this.status_text = this.__getInfo(status.get("text"));

        // Time details.
        this.time_details = this.__getDetails(flight_details.get("time"));

        // Flight trail.
        this.trail = flight_details.get("trail", list());
    }
}

module.exports = Flight;
